docs/ksvalidator.1
.TH "KSVALIDATOR" "1"
.SH "NAME"
ksvalidator \(em verify the syntax of a kickstart file
.SH "SYNOPSIS"
.PP
\fBksvalidator\fR [\fB\-e\fR | \fB\-\-firsterror\fP]  [\fB\-i\fR | \fB\-\-followincludes\fP]  [\fB\-l\fR | \fB\-\-listversions\fP]  [\fB\-v\fR | \fB\-\-version VERSION\fP]  INFILE [INFILE...]
.SH "DESCRIPTION"
.PP
\fBksvalidator\fR is a program that takes an input kickstart file and attempts to verify that it is syntactically correct.  This
includes checking basics like proper quoting and making sure required options have values, and more advanced checks like verifying
the file does not include deprecated or removed kickstart commands.  Mutually exclusive options can also be caught, as can certain
classes of partitioning errors (like using the same device name multiple times).
.PP
\fBksvalidator\fR does have its limitations.  It cannot run validate the syntax of %pre, %pre-install, and %post scripts, nor does it attempt to
run them.  It cannot check that the %packages section is valid.  Most importantly, it cannot guarantee that an input kickstart
file will install properly, because it does not understand the complexities of partitioning and what potentially already exists
on disk.
.SH "EXIT STATUS"
.PP
\fBksvalidator\fR returns 0 on success, and a value equal to amount of errors if some INFILEs do not exist or there are errors parsing the kickstart files.
.SH "OPTIONS"
.IP "\fB\-e\fP, \fB\-\-firsterror\fP" 10
Stop on the first warning or error.  By default, \fBksvalidator\fR will attempt to process the entire file, potentially raising
multiple errors.
.IP "\fB\-i\fP, \fB\-\-followincludes\fP" 10
Process included files in the input file when they are seen, like anaconda would.  If the included file does not exist (this can
happen for a variety of reasons, but the most common being a %pre script that generates an included file) and this option is
given, an error will be generated and processing halted.
.IP "\fB\-l\fP, \fB\-\-listversions\fP" 10
List all versions of kickstart syntax supported by \fBksvalidator\fR.  This is useful for determining what values can be
fed back in on a second run.
.IP "\fB\-v\fP, \fB\-\-version VERSION\fP" 10
Use this version of kickstart syntax when processing the file, or the latest if no version is given.
.SH "SEE ALSO"
.PP
ksflatten (1), ksshell (1), ksverdiff (1)


ChangeLog
2008-04-16  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py:  Bump version.

	* Makefile:  Update for git.

	* pykickstart/parser.py (KickstartHandlers.doNetwork): query is now a valid
	bootproto for networking (#441837).

2006-12-15  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py:  Bump version (only translations have
	changed since the last rebuild - #216620).

2006-12-05  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py:  Bump version.

2006-12-04  Chris Lumens  <clumens@redhat.com>

	* pykickstart/writer.py (KickstartWriter.doRepo):  Fix traceback
	(#218274).

2006-12-01  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py:  Bump version.

	* pykickstart/parser.py (KickstartHandlers.deprecatedCommands):  Fix
	traceback (#218047, #218059).

2006-11-30  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py:  Bump version.

2006-11-21  Chris Lumens  <clumens@redhat.com>

	* po/pykickstart.pot:  Updated (#216620).

	* pykickstart/parser.py:  Update string formats to make translation
	easier.

2006-11-14  Chris Lumens  <clumens@redhat.com>

	* pykickstart/parser.py (KickstartHandlers.doIscsi):  Use required
	attribute.

2006-11-09  Chris Lumens  <clumens@redhat.com>

	* pykickstart/writer.py (KickstartWriter.doLogging):  Add --level.

2006-10-24  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py:  Bump version.

	* Makefile:  Fix for new Release format.

2006-10-17  Chris Lumens  <clumens@redhat.com>

	* pykickstart/writer.py:  Add writer for key command.

2006-10-17  Jeremy Katz  <katzj@redhat.com>

	* pykickstart.spec, setup.py: Bump version.
	
	* pykickstart/parser.py, pykickstart/constants.py: add --skip for key

2006-10-16  Jeremy Katz  <katzj@redhat.com>

	* pykickstart.spec, setup.py: Bump version.

	* pykickstart/parser.py (KickstartHandlers.doKey): Support for
	installation key
	* pykickstart/data.py (KickstartData.__init__): Likewise.

2006-10-13  Bill Nottingham  <notting@redhat.com>

	* af.po, am.po, ar.po, be.po, bg.po, bn.po, cy.po, et.po,
	  eu_ES.po, fa.po, fi.po, he.po, hy.po, kn.po, ku.po, lo.po, lt.po,
	  lv.po, mk.po, my.po, nn.po, no.po, nso.po, ro.po, si.po, sq.po, th.po,
	  ur.po, vi.po, zu.po: Use a valid charset (#210720)
	
	* pykickstart.spec, setup.py: Bump version


2006-09-29  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py: Bump version

	* pykickstart/parser.py (KickstartHandlers.doMethod):  Fix traceback
	and test (#208557).

2006-09-25  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py: Bump version

	* pykickstart/parser.py (KickstartHandlers.doMethod):  Add support for
	--biospart option (#207585).
	* pykickstart/writer.py (KickstartWriter.doMethod):  Likewise.

	* pykickstart/writer.py (KickstartWriter.doIscsi):  Output new syntax.
	(KickstartWriter.doZFCP):  Likewise.
	(KickstartWriter.doIscsiName):  Added.

2006-09-20  Jeremy Katz  <katzj@redhat.com>

	* pykickstart.spec, setup.py: Bump version

	* pykickstart/parser.py (KickstartHandlers.doIscsi): New and
	improved iscsi syntax.  Also, allow multiple zfcp devs
	* pykickstart/data.py: Likewise.

2006-07-13  Chris Lumens  <clumens@redhat.com>

	* pykickstart/data.py:  Add noipv4 and noipv6 to KickstartNetworkData.
	* pykickstart/parser.py (KickstartHandlers.doNetwork):  Limit
	bootproto choices to what anaconda can handle.  Add support for
	--noipv4 and --noipv6 options.
	* pykickstart/writer.py (KickstartWriter.doNetwork):  Write out new
	networking options.

2006-06-20  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py:  Bump version.

	* pykickstart/writer.py (KickstartWriter.doIscsi):  Add iscsi writer.

2006-06-20  Jeremy Katz  <katzj@redhat.com>

	* pykickstart/parser.py (KickstartHandlers.doIscsi): Initial iscsi
	support.  This syntax is likely to change but this makes testing
	so much easier.
	* pykickstart/data.py (KickstartData.__init__): Add iscsi data.

2006-06-20  Chris Lumens  <clumens@redhat.com>

	* pykickstart/parser (KickstartHandlers.doRaid):  Handle a RAID device
	as an integer instead of a string (#176537).

2006-06-16  Jeremy Katz  <katzj@redhat.com>

	* pykickstart/parser.py (KickstartHandlers.doMethod): Handle nfs --opts

2006-06-06  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec:  Fix BuildRequires (#194156).

2006-05-25  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py:  Bump version.

2006-05-24  Chris Lumens  <clumens@redhat.com>

	* pykickstart/writer.py (KickstartWriter):  Change order of LVM-related
	functions (#193073).

2006-05-23  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec:  Require urlgrabber now.

	* po/*:  Update translation files.

	* pykickstart/parser.py (KickstartParser.handleCommand):  Return a
	more useful error message on unknown command.

	* pykickstart/writer.py (KickstartWriter.doLogicalVolume):  Fix typo.

	* validator/ksvalidator:  Accept a URL in addition to a file name.

2006-05-23  Chris Lumens  <clumens@redhat.com>

	* pykickstart/writer.py (KickstartWriter.doPackages):  Don't write out
	an empty packages section, since we may want this step to be
	interactive (#192851).

2006-05-23  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py:  Bump version.

	* pykickstart/writer.py (KickstartWriter.doFirewall):  Rename --ports
	to match the parser.

2006-05-17  Peter Jones  <pjones@redhat.com>

	* pykickstart/parser.py: Add "multipath" command and doMultiPath
	handler.
	* pykickstart/data.py: Add kickstartMpPathData and
	KickstartMultiPathData objects to handle multipath info, as well as
	KickstartData.mpaths .
	* pykickstart/writer.py: add doMultiPath
	* docs/programmers-guide: add completely useless documentation for
	multipath kickstart objects.

2006-05-15  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py:  Bump version.

	* pykickstart/data.py (KickstartNetworkData):  Add MTU (#191328).

tools/ksvalidator.py
#!/usr/bin/env python3
#
# Chris Lumens <clumens@redhat.com>
#
# Copyright 2005-2014 Red Hat, Inc.
#
# This copyrighted material is made available to anyone wishing to use, modify,
# copy, or redistribute it subject to the terms and conditions of the GNU
# General Public License v.2.  This program is distributed in the hope that it
# will be useful, but WITHOUT ANY WARRANTY expressed or implied, including the
# implied warranties of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc., 51
# Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.  Any Red Hat
# trademarks that are incorporated in the source code or documentation are not
# subject to the GNU General Public License and may only be used or replicated
# with the express permission of Red Hat, Inc.
#

# Disable found-_-in-module-class
# pylint: disable=W9902

import argparse
import os
import sys
import warnings
import tempfile
import shutil
from pykickstart.i18n import _
from pykickstart.errors import KickstartError, KickstartParseError, KickstartVersionError,\
    KickstartParseWarning, KickstartDeprecationWarning
from pykickstart.load import load_to_file
from pykickstart.parser import KickstartParser, preprocessKickstart
from pykickstart.version import DEVEL, makeVersion, versionMap

def cleanup(dest, fn=None, exitval=1):
    shutil.rmtree(dest)

    # Don't care if this file doesn't exist.
    if fn is not None:
        try:
            os.remove(fn)
        except FileNotFoundError:
            pass

    return exitval

def main(argv):
    op = argparse.ArgumentParser(usage="%(prog)s [options] ksfile", add_help=False)
    op.add_argument("ksfile", nargs="?",
                    help=_("filename or URL to read from"))
    op.add_argument("-e", "--firsterror", dest="firsterror", action="store_true",
                    default=False, help=_("halt after the first error or warning"))
    op.add_argument("-i", "--followincludes", dest="followincludes",
                    action="store_true", default=False,
                    help=_("parse include files when %%include is seen"))
    op.add_argument("-l", "--listversions", dest="listversions", action="store_true",
                    default=False,
                    help=_("list the available versions of kickstart syntax"))
    op.add_argument("-v", "--version", dest="version", default=DEVEL,
                    help=_("version of kickstart syntax to validate against"))
    op.add_argument("-h", "--help", dest="help", action="store_true", default=False,
                    help=_("show this help message and exit"))

    opts = op.parse_args(argv)

    # parse --help manually b/c we don't want to sys.exit before the
    # tests have finished
    if opts.help:
        return (0, op.format_help().split("\n"))

    if opts.listversions:
        versions = []
        for key in sorted(versionMap.keys()):
            versions.append(key)
        return (0, versions)

    if not opts.ksfile:
        return (1, op.format_usage().split("\n"))

    destdir = tempfile.mkdtemp(prefix="ksvalidator-tmp-")
    try:
        f = load_to_file(opts.ksfile, os.path.join(destdir, "ks.cfg"))
    except KickstartError as e:
        return (cleanup(destdir),
                [_("Error reading %(filename)s:\n%(version)s") % {"filename": opts.ksfile, "version": e}])

    try:
        handler = makeVersion(opts.version)
    except KickstartVersionError:
        return (cleanup(destdir),
                [_("The version %s is not supported by pykickstart") % opts.version])

    ksparser = KickstartParser(handler, followIncludes=opts.followincludes,
                               errorsAreFatal=opts.firsterror)

    # turn kickstart parse warnings into errors
    warnings.filterwarnings(action="error", category=KickstartParseWarning)

    processedFile = None

    try:
        processedFile = preprocessKickstart(f)
        if processedFile is None:
            raise RuntimeError("Empty file")
        ksparser.readKickstart(processedFile)
        return (cleanup(destdir, processedFile, exitval=ksparser.errorsCount), [])
    except KickstartDeprecationWarning as err:
        return (cleanup(destdir, processedFile),
                [_("File uses a deprecated option or command.\n%s") % err])
    except KickstartParseError as err:
        return (cleanup(destdir, processedFile), [str(err)])
    except KickstartError:
        return (cleanup(destdir, processedFile),
                [_("General kickstart error in input file")])
    except Exception as e:                                      # pylint: disable=broad-except
        return (cleanup(destdir, processedFile),
                [_("General error in input file:  %s") % e])

if __name__ == "__main__":
    retval, messages = main(sys.argv[1:])
    for msg in messages:
        print(msg)
    sys.exit(retval)
